influence_benchmark/RL/expert_iteration.py

```diff
@@ -0,0 +1,217 @@
+            for i, done in enumerate(is_done_n):
+                if done:
+                    trajectory_ids[i] = next_trajectory_id
+                    next_trajectory_id += 1
+
+            observations = vec_env.get_observation_vec()
+            actions = agent.get_action_vec(observations)
+            next_states, _ = vec_env.step_vec(actions)
+            observations = vec_env.get_observation_vec()
+
+            for i, state in enumerate(next_states):
+                env_trajectories[i].append(
+                    {
+                        "trajectory_id": trajectory_ids[i],
+                        "env_id": i,
+                        "turn": state.turns,
+                        "agent_system_prompt": agent.get_system_prompt(observations[i]),
+                        "history": state.history[:-1],
+                        "preferences": state.preferences,
+                        "transition_probs": state.transition_probs,
+                    }
+                )
+
+        env_trajectories = [
+            [traj for traj in trajectories if int(traj["trajectory_id"]) < start_trajectory_id + num_trajectories]
+            for trajectories in env_trajectories
+        ]
+
+        save_path = trajectory_folder / f"{device.split(':')[-1]}.jsonl"
+        save_path.parent.mkdir(parents=True, exist_ok=True)
+        with open(save_path, "w", encoding="utf-8") as f:
+            for env in env_trajectories:
+                for turn_data in env:
+                    f.write(json.dumps(turn_data) + "\n")
+
+        backend.close()
+
+    def rank_trajectories_by_avg_reward(self, trajectory_folder):
+        trajectories = []
+        for file in trajectory_folder.iterdir():
+            if file.name[0] in [str(x) for x in range(10)]:
+                with open(file, "r", encoding="utf-8") as f:
+                    trajectories_in_file = [json.loads(line) for line in f]
+                    trajectories.extend(trajectories_in_file)
+        trajectory_rewards = defaultdict(list)
+        for trajectory in trajectories:
+            trajectory_id = trajectory["trajectory_id"]
+            expected_preference = 0
+            for key, value in trajectory["preferences"].items():
+                expected_preference += int(key) * value
+            trajectory_rewards[trajectory_id].append(expected_preference)
+
+        avg_rewards = {key: sum(value) / len(value) for key, value in trajectory_rewards.items()}
+        sorted_trajectories = sorted(trajectories, key=lambda x: avg_rewards[x["trajectory_id"]], reverse=True)
+        num_selected = 0
+        selected_trajectories = []
+        selected_trajectory_ids = set()
+        for trajectory in sorted_trajectories:
+            selected_trajectories.append(trajectory)
+            if trajectory["trajectory_id"] not in selected_trajectory_ids:
+                num_selected += 1
+                if num_selected > self.num_chosen_trajectories:
```

Comment:
This is across many possible starting states? It may be useful to do selection as BoN for some N across the same starting state